# Variables
* we can use python like a calculator

3 + 5 * 4

* or we can assign things to variables

In [63]:
weight_kg = 60

* weight_kg is variable name
* variable names are case sensitive, can include letters, digits and underscores
* can't start with a digit

* python has different data types
* integers, floats, strings
* automatically determined, 60 is an integer, 60.3 is a float

In [64]:
weight_kg = 60.3

* strings are surrouneded by single or double quotes

In [65]:
patient_id = '001'

* we can store the results of calculations in a variable

In [66]:
weight_lb = 2.2 * weight_kg

* we can add a prefix to our patient id, this joins two strings together

In [67]:
patient_id = 'inflam_' + patient_id

* functions do something with some data
* lots of built in functions like print
* brackets around parameters to the function

In [68]:
print(weight_lb)

132.66


In [69]:
print(patient_id)

inflam_001


* display multiple things in one call to print by using commas
* automatically adds spaces

In [70]:
print(patient_id, 'weight in kilos:', weight_kg)

inflam_001 weight in kilos: 60.3


* function calls can be nested
* type function tells us what type a variable is

In [71]:
print(type(60.3))

<class 'float'>


In [72]:
print(type(patient_id))

<class 'str'>


* we can do arithmetic inside a function call

In [73]:
print('weight in pounds:', 2.2 * weight_kg)

weight in pounds: 132.66


In [74]:
print(weight_kg)

60.3


* we didn't change weight_kg
* have to assign a new value with = 

In [75]:
weight_kg = 65.0
print('weight in kilos is now:', weight_kg)

weight in kilos is now: 65.0


In [76]:
weight_lb = 2.2 * weight_kg
print('weight in kilos:', weight_kg, 'and in pounds:', weight_lb)

weight in kilos: 65.0 and in pounds: 143.0


In [77]:
weight_kg = 100.0
print('weight in kilos is now:', weight_kg, 'and weight in pounds is still:', weight_lb)

weight in kilos is now: 100.0 and weight in pounds is still: 143.0


* think of variables like sticky notes
* each has a name and value
* changing one doesn't change others

## Exercises

In [78]:
mass = 47.5
age = 122
mass = mass * 2.0
age = age - 20



In [79]:
first, second = 'Grace', 'Hopper'
third, fourth = second, first
print(third, fourth)



Hopper Grace


In [80]:
planet = 'Earth'
apples = 5
distance = 10.5



## recap

* Basic data types in Python include integers, strings, and floating-point numbers.
* Use variable = value to assign a value to a variable in order to record it in memory.
* Variables are created on demand whenever a value is assigned to them.
* Use print(something) to display the value of something.
* Built-in functions are always available to use.


# Analysing some wave-height data

In [81]:
import numpy

* explain libraries, like taking equipment out of storage.
* use import statement
* libraryname.function name syntax

In [82]:
numpy.loadtxt(fname='data/wavesmonthly.csv', delimiter=',', skiprows=1)


array([[1.979e+03, 1.000e+00, 3.788e+00],
       [1.979e+03, 2.000e+00, 3.768e+00],
       [1.979e+03, 3.000e+00, 4.774e+00],
       ...,
       [2.015e+03, 1.000e+01, 3.046e+00],
       [2.015e+03, 1.100e+01, 4.622e+00],
       [2.015e+03, 1.200e+01, 5.048e+00]])

* data now loaded, only subset onscreen
* explain parameters
* skipped first row since it contains headers
* we didn't save the data to a variable, correct that with variable assignment

In [83]:
data = numpy.loadtxt(fname='data/wavesmonthly.csv', delimiter=',', skiprows=1)

In [84]:
print(data)

[[1.979e+03 1.000e+00 3.788e+00]
 [1.979e+03 2.000e+00 3.768e+00]
 [1.979e+03 3.000e+00 4.774e+00]
 ...
 [2.015e+03 1.000e+01 3.046e+00]
 [2.015e+03 1.100e+01 4.622e+00]
 [2.015e+03 1.200e+01 5.048e+00]]


In [85]:
print(type(data))

<class 'numpy.ndarray'>


* numpy arrays can contain multiple data types
* the type function tells us the type of a variable
* dtype tells us the type numpy is using within an array


In [86]:
print(data.dtype)

float64


* find out the array's shape

In [87]:
print(data.shape)

(444, 3)


* that's 444 rows, 3 columns
* 37 years x 12 months = 444
* 3 columns are year, month, waveheight
* get data from an array with [x,y]

In [88]:
print('first value in data:', data[0, 2])

first value in data: 3.788


In [89]:
print("middle value in data:", data[222, 2])

middle value in data: 2.446


* show diagram in notes about arrays
* arrays are zero based indexing
* 0,0 is top left corner, matrix notation not cartesian

## slicing data
* can select multiple items in an array with a :

In [90]:
print(data[0:12, 0:3])

[[1.979e+03 1.000e+00 3.788e+00]
 [1.979e+03 2.000e+00 3.768e+00]
 [1.979e+03 3.000e+00 4.774e+00]
 [1.979e+03 4.000e+00 2.818e+00]
 [1.979e+03 5.000e+00 2.734e+00]
 [1.979e+03 6.000e+00 2.086e+00]
 [1.979e+03 7.000e+00 2.066e+00]
 [1.979e+03 8.000e+00 2.236e+00]
 [1.979e+03 9.000e+00 3.322e+00]
 [1.979e+03 1.000e+01 3.512e+00]
 [1.979e+03 1.100e+01 4.348e+00]
 [1.979e+03 1.200e+01 4.628e+00]]


* 0:12 means from zero to 11
* takes some getting used to, I get it wrong all the time
* slices can start from non-zero positions

In [91]:
print(data[12:24, 1:3])

[[ 1.     3.666]
 [ 2.     4.326]
 [ 3.     3.522]
 [ 4.     3.18 ]
 [ 5.     1.954]
 [ 6.     1.72 ]
 [ 7.     1.86 ]
 [ 8.     1.95 ]
 [ 9.     3.11 ]
 [10.     3.78 ]
 [11.     3.474]
 [12.     5.28 ]]


* we can miss out the start or end if we want everything up to or everything after an item

In [92]:
first_year = data[:12, 2:]
print('data from first year is:')
print(first_year)


data from first year is:
[[3.788]
 [3.768]
 [4.774]
 [2.818]
 [2.734]
 [2.086]
 [2.066]
 [2.236]
 [3.322]
 [3.512]
 [4.348]
 [4.628]]


* numpy has functions which can operate on whole arrays
* mean is one of these

In [93]:
print(numpy.mean(data))

668.9611876876877


* mean takes an argument (data)
* we can also slice data to just get the mean of a subset of the data

In [94]:
print(numpy.mean(data[:,2]))

3.383563063063063


* Not all functions take an argument

In [95]:
import time
print(time.ctime())



Wed May 17 14:05:18 2023


* ctime from the time library needs no arguments, still need the brackets
* lets combine multiple assignment, numpy's min/max/std functions and slicing

In [96]:
maxval, minval, stdval = numpy.max(data[:,2]), numpy.min(data[:,2]), numpy.std(data[:,2])

print('Max wave height:', maxval)
print('Min wave height:', minval)
print('Wave height standard deviation:', stdval)

Max wave height: 6.956
Min wave height: 1.496
Wave height standard deviation: 1.1440155050316319


* don't use variable names with the same name as functions (e.g. min/max)
* mystery functions

** Potential issue **

The what about NaNs section creates a new data variable, which overwrites the one we previously had.

To avoid running into problems when doing the reshaping that comes next, what about modifying the name of the 'data' var defined here.

In [97]:
data_nan = numpy.array([[1,2,3], [1, numpy.NaN, 3], [1, 2, 3]])
numpy.mean(data_nan)

nan

In [98]:
data_nan = numpy.array([[1,2,3],[1,numpy.NaN,3],[1,2,3]])
numpy.nanmean(data_nan)

2.0

In [99]:
data_nan.shape

(3, 3)

In [100]:
reshaped_data = numpy.reshape(data[:,2], [37,12])

In [102]:
print("The shape of the reshaped data is:")
print(reshaped_data.shape)

The shape of the reshaped data is:
(37, 12)


In [104]:
print(f"The maximum value of the reshaped data is: {numpy.max(reshaped_data)}")
print(f"The min value of the reshaped data is: {numpy.min(reshaped_data)}")
print(f"The std dev of the reshaped data is: {numpy.std(reshaped_data)}")

The maximum value of the reshaped data is: 6.956
The min value of the reshaped data is: 1.496
The std dev of the reshaped data is: 1.1440155050316319


**ISSUE**
After the line "We can verify that nothing about the data has changed:"

the name of the variable should be reshaped_data and not reshaped

### Printing text

In [107]:
name = 'Chris'
city = 'Edinburgh'
print(f"Hello, My name is {name} and I live in {city}.")

Hello, My name is Chris and I live in Edinburgh.


In [108]:
year_0 = reshaped_data[0,:] # 0 on the first axis (rows), everything on the second (columns)
print(f"maximum wave height for year 0: {numpy.max(year_0)}")

maximum wave height for year 0: 4.774


In [109]:
print(numpy.mean(reshaped_data, axis=0))

[4.59956757 4.39708108 4.09156757 3.26016216 2.60437838 2.3072973
 2.18940541 2.32145946 2.9907027  3.55627027 3.90345946 4.38140541]


In [110]:
print(numpy.mean(reshaped_data, axis=0).shape)

(12,)


**Potential issue**

Technically we don't have a Nx1 vector. A vector is a Nx1 matrix by definition, but what we have here is a 1D array. (12,) is different from (12,1) becasue the former is 1D and the latter is 2D. This is very important when you are doing matrix multiplications because the (12,) version is misleading and often leads to errors. Not sure if we just ignore this for now?

In [111]:
print(numpy.mean(reshaped_data, axis=1))

[3.34       3.15183333 3.29866667 3.53366667 3.448      3.23016667
 2.99383333 3.51133333 2.96066667 3.20316667 3.62116667 5.1915
 3.28816667 3.529      3.523      3.66866667 3.314      2.99916667
 3.45983333 3.16783333 3.413      3.3435     3.031      3.29366667
 3.138      3.29716667 3.3185     3.24966667 3.4135     3.42866667
 3.168      2.78816667 3.61366667 3.2725     3.32766667 3.2765
 4.385     ]


# Slicing Strings

In [112]:
element = 'oxygen'
print('first three characters:', element[0:3])
print('last three characters:', element[3:6])

first three characters: oxy
last three characters: gen


In [114]:
element = 'oxygen'
print('last three characters:', element[-3:])
element = 'carpentry'
print('last three characters:', element[-3:])
element = 'clone'
print('last three characters:', element[-3:])
element = 'hi'
print('last three characters:', element[-3:])

last three characters: gen
last three characters: try
last three characters: one
last three characters: hi


In [120]:
data[3:3, 4:4]

reshaped_data.shape

(37, 12)

In [118]:
data.shape

(444, 3)

### Stacking arrays

In [121]:
A = numpy.array([[1,2,3], [4,5,6], [7, 8, 9]])
print('A = ') 
print(A)  
B = numpy.hstack([A, A]) 
print('B = ') 
print(B)  
C = numpy.vstack([A, A]) 
print('C = ') 
print(C)

A = 
[[1 2 3]
 [4 5 6]
 [7 8 9]]
B = 
[[1 2 3 1 2 3]
 [4 5 6 4 5 6]
 [7 8 9 7 8 9]]
C = 
[[1 2 3]
 [4 5 6]
 [7 8 9]
 [1 2 3]
 [4 5 6]
 [7 8 9]]


In [122]:
D = numpy.hstack((A[:, :1], A[:, -1:]))
print('D = ')
print(D)

D = 
[[1 3]
 [4 6]
 [7 9]]


In [128]:
D = numpy.delete(A, 1, 1)
print('D = ')
print(D) 

D = 
[[1 3]
 [4 6]
 [7 9]]


### Change In Wave Height

In [136]:
year4 = reshaped_data[3, :]
print(year4)


[3.73  4.886 4.76  3.188 2.528 1.662 1.952 2.388 3.336 4.034 4.502 5.438]


In [137]:
numpy.diff(year4)

array([ 1.156, -0.126, -1.572, -0.66 , -0.866,  0.29 ,  0.436,  0.948,
        0.698,  0.468,  0.936])

In [138]:
numpy.max(numpy.diff(reshaped_data, axis=1), axis=1)

array([1.086, 1.806, 1.776, 1.156, 1.692, 1.274, 0.798, 2.59 , 1.338,
       1.634, 0.992, 0.618, 1.054, 1.652, 1.472, 1.716, 0.766, 1.496,
       1.656, 1.04 , 1.228, 1.336, 1.564, 1.066, 1.242, 1.604, 0.802,
       1.04 , 0.652, 0.86 , 1.176, 0.97 , 1.68 , 1.556, 1.904, 2.936,
       1.578])

In [139]:
numpy.max(numpy.absolute(numpy.diff(reshaped_data, axis=1)), axis=1)

array([1.956, 1.806, 1.776, 1.572, 3.6  , 2.418, 0.954, 2.798, 1.338,
       1.634, 2.13 , 0.93 , 1.054, 1.71 , 1.68 , 2.   , 1.614, 1.496,
       2.308, 1.04 , 2.014, 1.68 , 1.564, 1.596, 1.528, 1.604, 1.468,
       1.21 , 1.012, 0.86 , 1.732, 1.03 , 1.68 , 1.774, 1.904, 2.936,
       1.694])

(37, 11)